### PR TITLE
Support custom logits processors

### PR DESCRIPTION
**Pull request by AllentDan**
```python
from lmdeploy import pipeline
from lmdeploy.messages import GenerationConfig, PytorchEngineConfig

model = 'InternLM/internlm2-chat-1_8b'

pipe = pipeline(model, backend_config=PytorchEngineConfig(), log_level='INFO')
def process_token(token_ids, logits):
    logits[6623] = -9999.999 
    logits[4853] = -9999.999
    logits[630] = -9999.999
    logits[985] = -9999.999
    return logits

gen_config = GenerationConfig(logits_processors=[process_token])

# response = pipe(('describe the image', image))
response = pipe(['Make a self introduction please.', 'hi']*2, gen_config=gen_config)
print(response)
```

### CODE REVIEW

**Comment by lvhan028** [comment]
We need to support `logits_processors` in api_server, too

File: lmdeploy/pytorch/engine/engine.py [hunk 1]
```diff
@@ -492,9 +492,32 @@ def __get_last_logits():
...
+                            logits_cloned[seq_id] = processor(
+                                input_id, logits_cloned[seq_id])
+                    else:
+                        for processor in processors:
+                            logits_cloned[seq_id] = processor(
+                                history_ids[seq_id], logits_cloned[seq_id])
+            return logits_cloned
+
         split_logits = __get_last_logits().cuda()
+        logits = _apply_logits_processors(sampling_inputs.logits_processors,
```
Comment (grimoire) [inline lmdeploy/pytorch/engine/engine.py:516]: hide it in logits proecessor

**Review by RunningLeon — APPROVED** [review]
LGTM

File: lmdeploy/pytorch/engine/logits_process.py [hunk 2]
```diff
@@ -216,6 +222,24 @@ def to_device(self, device: str):
...
+        if processors is not None:
+            if history_ids.shape[-1] == 0:
+                seq_cumsum = [0] + seq_length.cumsum(0).tolist()
+                input_id = input_ids[0][seq_cumsum[seq_id]:seq_cumsum[seq_id +
+                                                                      1]]
+                for processor in processors:
+                    logits[seq_id] = processor(input_id, logits[seq_id])
+            else:
+                for processor in processors:
+                    logits[seq_id] = processor(history_ids[seq_id],
```
Comment (grimoire) [inline lmdeploy/pytorch/engine/logits_process.py:238]: what is the first input, input_ids or history_ids?

Comment (AllentDan) [reply lmdeploy/pytorch/engine/logits_process.py:238]: It's just `function(previous_token_ids, logits)`. I wanted to align with vllm.
https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/logits_processor.py#L137

But I must pass more arguments in if I want to get `prompt_tokens_ids` and `past_tokens_ids`.

Comment (grimoire) [reply lmdeploy/pytorch/engine/logits_process.py:238]: Should we change history_ids to all_ids? That would also benefit replicate penalty.

**Review by grimoire — APPROVED** [review]

File: lmdeploy/pytorch/engine/engine.py [hunk 3]
```diff
@@ -494,7 +494,7 @@ def __get_last_logits():
 
         split_logits = __get_last_logits().cuda()
         logits_processor = FusedLogitsProcessor(sampling_inputs, ignore_eos)
-        logits = logits_processor(history_ids, split_logits)
+        logits = logits_processor(split_logits, all_ids)
```
Comment (lvhan028) [inline lmdeploy/pytorch/engine/engine.py:497]: (split_logits, all_ids) or (all_ids, split_logits)?

File: lmdeploy/pytorch/engine/logits_process.py [hunk 4]
```diff
@@ -224,17 +240,17 @@ def __init__(self, sampling_inputs: SamplingInputs,
         self.sampling_inputs: SamplingInputs = sampling_inputs
         self.ignore_eos = ignore_eos
 
-    def __call__(self, input_ids: torch.LongTensor,
-                 scores: torch.FloatTensor) -> torch.FloatTensor:
+    def __call__(self, scores: torch.FloatTensor,
```
Comment (lvhan028) [inline lmdeploy/pytorch/engine/logits_process.py:243]: why change the sequence of scores and all_ids

File: lmdeploy/pytorch/engine/logits_process.py [hunk 5]
```diff
@@ -224,17 +240,17 @@ def __init__(self, sampling_inputs: SamplingInputs,
...
         r"""
         Args:
-            input_ids (torch.LongTensor):
-                Indices of input sequence tokens in the vocabulary.
             scores (torch.FloatTensor):
                 Prediction scores of a language modeling head.
                 These can be logits for each vocabulary when not using
                 beam search or log softmax for each vocabulary token
                 when using beam search
+            all_ids (torch.FloatTensor): All the token ids.
```
Comment (lvhan028) [inline lmdeploy/pytorch/engine/logits_process.py:252]: Is all_ids `torch.FloatTensor`?

File: lmdeploy/pytorch/engine/engine.py [hunk 6]
```diff
@@ -478,7 +478,7 @@ def _batch_stopping_criteria(self, token_ids: torch.Tensor,
 
     @logging_timer('SamplingLogits', logger)
     def async_sampling_logits(self, logits: torch.Tensor,
-                              history_ids: torch.Tensor,
+                              all_ids: torch.Tensor,
```
Comment (lvhan028) [inline lmdeploy/pytorch/engine/engine.py:481]: The changes in this file are mainly about renaming "history" to "all". 
What's the motivation?

Comment (AllentDan) [reply lmdeploy/pytorch/engine/engine.py:481]: History_ids does not contain prompt_token_ids in the first iteration of genration. This is for both `repetition_panelty` and `custom_logits_processors`.

File: lmdeploy/pytorch/engine/logits_process.py (hunk 4 above)
Comment (AllentDan) [reply lmdeploy/pytorch/engine/logits_process.py:243]: I did not change it, I just delete input_ids and put all_ids.
